### PR TITLE
Improvements to user defined timestamp

### PR DESCRIPTION
**Pull request by larry0x**
Closes: #904 

1. Fixes a memory bug in the following methods:
  - `ReadOptions::set_timestamp`
  - `ReadOptions::set_iter_start_ts`
  - `CompactOptions::set_full_history_ts_low`
2. Modify the `Options::set_comparator` API, allowing users to choose their own timestamp and comparator implementations (unlike in #901, which forces using big endian u64).

### CODE REVIEW

File: src/comparator.rs [hunk 1]
```diff
@@ -15,164 +15,103 @@
...
-pub unsafe extern "C" fn name_callback(raw_cb: *mut c_void) -> *const c_char {
-    let cb: &mut ComparatorCallback = &mut *(raw_cb as *mut ComparatorCallback);
-    let ptr = cb.name.as_ptr();
-    ptr as *const c_char
+pub struct ComparatorWithTsCallback {
+    pub name: CString,
+    pub compare_fn: Box<CompareFn>,
+    pub compare_ts_fn: Box<CompareTsFn>,
+    pub compare_without_ts_fn: Box<CompareWithoutTsFn>,
 }
```
Comment (larry0x) [inline src/comparator.rs:62]: Now we have two structs here:

- `ComparatorCallback`, which is used in `Options::set_comparator`
- `ComparatorWithTsCallback`, which is used in `Options::set_comparator_with_ts`

File: src/comparator.rs [hunk 2]
```diff
@@ -15,164 +15,103 @@
...
-    b_raw: *const c_char,
-    b_len: size_t,
-) -> c_int {
-    let cb: &mut ComparatorCallback = &mut *(raw_cb as *mut ComparatorCallback);
-    let a: &[u8] = slice::from_raw_parts(a_raw as *const u8, a_len);
-    let b: &[u8] = slice::from_raw_parts(b_raw as *const u8, b_len);
-    match (cb.f)(a, b) {
-        Ordering::Less => -1,
-        Ordering::Equal => 0,
-        Ordering::Greater => 1,
```
Comment (larry0x) [inline src/comparator.rs:81]: This match statement is unnecessary. We can simply `.into()`

File: src/db_options.rs [hunk 3]
```diff
@@ -342,6 +344,12 @@ pub struct BlockBasedOptions {
...
 pub struct ReadOptions {
     pub(crate) inner: *mut ffi::rocksdb_readoptions_t,
+    // The `ReadOptions` owns a copy of the timestamp and iteration bounds.
+    // This is necessary to ensure the pointers we pass over the FFI live as
+    // long as the `ReadOptions`. This way, when performing the read operation,
+    // the pointers are guaranteed to be valid.
+    timestamp: Option<Vec<u8>>,
+    iter_start_ts: Option<Vec<u8>>,
     iterate_upper_bound: Option<Vec<u8>>,
     iterate_lower_bound: Option<Vec<u8>>,
```
Comment (larry0x) [inline src/db_options.rs:354]: This is the centerpiece of the memory bug fix

File: src/db_options.rs [hunk 4]
```diff
@@ -3872,26 +3888,43 @@ impl ReadOptions {
...
+        };
+        self.timestamp = ts;
         unsafe {
-            ffi::rocksdb_readoptions_set_timestamp(
-                self.inner,
-                ts.as_ptr() as *const c_char,
-                ts.len() as size_t,
-            );
+            ffi::rocksdb_readoptions_set_timestamp(self.inner, ptr, len);
         }
```
Comment (larry0x) [inline src/db_options.rs:3909]: We use exactly the same approach as in `ReadOptions::set_iterator_{lower,upper}_bound`

**Comment by zaidoon1** [comment]
can you address the clippy failures, once the builds pass, this looks good to me overall

**Comment by larry0x** [comment]
> can you address the clippy failures, once the builds pass, this looks good to me overall

Done

**Comment by zaidoon1** [comment]
there is build failures, can you please address them

**Review by zaidoon1 — APPROVED** [review]

**Comment by larry0x** [comment]
Sorry @zaidoon1, I pushed a new commit: 97eae778efe925c349e056ff7eb956bc36e4f3b2. Can you re-run the CI?

**Review by stanislav-tkach — APPROVED** [review]
LGTM. 👍

**Comment by zaidoon1** [comment]
@larry0x can you squash your commits/rebase so I can merge

**Pull request merged**